### ISSUE TITLE
Align Ditto Plugin with Documentation

### ISSUE THREAD

**Issue by kolotu** [eclipse/vorto#2568]
Align the implementation with the docs, if there are deviation. 
https://docs.bosch-iot-suite.com/asset-communication/Specification-digital-twin-model-mapping.html

**Comment by thjaeckle** [comment]
I think mainly the generation of schemas for operations and events has to be looked at and verified.
And the `dictionary` type is - as far as I remember - also not yet supported by the generator, that one should be added.

**Comment by mena-bosch** [comment]
I've browsed through the linked documentation page and tested against Vorto's Ditto generator for a few models (including the MightyLamp). 
So far, no surprises. 
Schemas are generated both for operations and events, data type mapping looks good, and the generator seems to do its job. 

I have observed a few minor issue with the documentation such as formatting in some cases (e.g. no whitespace between "readable", "writable" and "eventable" in the mapping overview. 
Also the `<namespace>` notation could actually be completed with `<namespace(.subnamespace)*?>` but as discussed with @kolotu this has bigger chances to confuse the reader rather than help them, so it's not worth addressing. 

I have not tested any runtime data (i.e. thing instances - which is amply documented in the page) since that would be tantamount to test Ditto rather than Vorto. 

**TL;DR**
Looks ok to me.

**Comment by thjaeckle** [comment]
A few things coming to my mind which do not work as specified:
* [events with only 1 parameter](https://docs.bosch-iot-suite.com/asset-communication/Specification-digital-twin-model-mapping.html#src-1352239526_Specificationdigitaltwinmodelmapping-Digitaltwinfacetevents): "When exactly 1 property is defined, the property's type (either primitive type, entity type or enum type) becomes the complete payload of the message." - the generator does not do that AFAIR
   * as far as I remember the generator does this for operations with 1 param, but not for events with 1 param
* and as mentioned above - Vorto `dictionary` is not supported at all by the generator (for JSON Schema)

**Comment by mena-bosch** [comment]
@thjaeckle thanks for that - reopening to check. 
For the dictionary, I have not looked at the JSON schema so there might be something there, although it seems supported otherwise.

**Comment by thjaeckle** [comment]
The issue is only about the JSON Schema I think - the JSON thing examples were added recently, the JSON Schema generation is a few years old.

**Comment by mena-bosch** [comment]
Hi @thjaeckle , I had to take a break from this one and I'm getting back to it now. 
I just re-tested the dictionary contextually to schema generation and as far as I can tell it is generated (as type "object"). 
Did you have any specific case in mind that I'm not seeing?

Will be looking at the 1 operation vs 1 event coalescing behavior shortly.

**Comment by thjaeckle** [comment]
@mena-bosch ah, there probably is a fallback to Json object if the type is not primitive or an entity.

That could be sufficient for the Json schema.
It could still be improved, because the value of the dictionary map could be enforced in json schema.
But as long as nobody needs that from the generator, it's probably fine the way it is.

**Comment by mena-bosch** [comment]
Thanks for your response @thjaeckle .
Wrt event vs operation: I have now compared the JSON schema for a 1-property event with the equivalent for an operation taking 1 parameter. 
Both coalesce the return type respectively as the single property of the event, and the type of the operation's single parameter.
Am I misinterpreting your comment once again somehow? Thanks for bearing with me here.

**Comment by thjaeckle** [comment]
Oh wow, I really didn't remember that and was sure that for events with a single property the property's value is warped in its property name.
Perfect if that is not the case, and sorry to have caused you so much trouble looking into it.

**Comment by mena-bosch** [comment]
@thjaeckle no trouble at all! 
Glad we could align as this area is not my speciality. 
I think it's safe to close now then. 
If you think of anything else, don't hesitate to give me a shout though.
